File: LeetCode/1190. Reverse Substrings Between Each Pair of Parentheses/1190.java
class Solution {
    public String reverseParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        int n = s.length();
        int[] pair = new int[n];
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else if (s.charAt(i) == ')') {
                int j = stack.pop();
                pair[i] = j;
                pair[j] = i;
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0, d = 1; i < n; i += d) {
            if (s.charAt(i) == '(' || s.charAt(i) == ')') {
                i = pair[i];
                d = -d;
            } else {
                sb.append(s.charAt(i));
            }
        }
        return sb.toString();

    }
}

/*
 * class Solution {
 * int i = 0;
 * public String reverseParentheses(String s) {
 * char[] ar = s.toCharArray();
 * return helper(ar);
 * }
 * 
 * public String helper(char[] s){
 * StringBuilder sb = new StringBuilder();
 * 
 * while(i < s.length){
 * if(s[i] == ')'){
 * i++;
 * return sb.reverse().toString();
 * }else if(s[i] == '('){
 * i++;
 * String st = helper(s);
 * //System.out.println(st);
 * sb.append(st);
 * }else{
 * sb.append(s[i]);
 * i++;
 * }
 * }
 * return sb.toString();
 * 
 * }
 * }
 */

/*
 * class Solution {
 * public String reverseParentheses(String s) {
 * int n=s.length();
 * Stack<StringBuilder> stack = new Stack<>();
 * StringBuilder str = new StringBuilder();
 * for(int i=0;i<n;i++){
 * char ch = s.charAt(i);
 * if(ch=='('){
 * stack.push(str);
 * str=new StringBuilder();
 * }
 * else if(ch==')'){
 * str.reverse();
 * if(!stack.isEmpty()){
 * stack.peek().append(str);
 * str=stack.pop();
 * }
 * }
 * else str.append(ch);
 * }
 * return str.toString();
 * }
 * }
 */
/*
 * class Solution {
 * public String reverseParentheses(String s) {
 * Stack<StringBuilder> st=new Stack<>();
 * StringBuilder str=new StringBuilder("");
 * for(int i=0;i<s.length();i++){
 * if(s.charAt(i)=='('){
 * st.push(str);
 * str=new StringBuilder("");
 * }
 * else if(s.charAt(i)==')'){
 * str.reverse();
 * StringBuilder str2=st.pop();
 * str2.append(str);
 * str=str2;
 * 
 * }
 * else{
 * str.append(s.charAt(i));
 * }
 * }
 * return str.toString();
 * }
 * }
 */

File: LeetCode/1190. Reverse Substrings Between Each Pair of Parentheses/1190.cs
public class Solution
{
    public string ReverseParentheses(string s)
    {
        Stack<int> stack = new Stack<int>();
        int n = s.Length;
        int[] pair = new int[n];
        for (int i = 0; i < n; i++)
        {
            if (s[i] == '(')
            {
                stack.Push(i);
            }
            else if (s[i] == ')')
            {
                int j = stack.Pop();
                pair[i] = j;
                pair[j] = i;
            }
        }
        StringBuilder sb = new StringBuilder();
        int index = 0, step = 1;
        while (index < n)
        {
            if (s[index] == '(' || s[index] == ')')
            {
                index = pair[index];
                step = -step;
            }
            else
            {
                sb.Append(s[index]);
            }
            index += step;
        }
        return sb.ToString();

    }
}

/
/*
public class Solution {
    public string ReverseParentheses(string s) {
        Stack<char> box = new Stack<char>();
        Queue<char> box2= new Queue<char>();
        foreach(var c in s)
        {
            if(c == ')')
            {
                while(box.Count != 0 && box.Peek() != '(') 
                    box2.Enqueue(box.Pop());
                box.Pop();
                while(box2.Count!= 0) box.Push(box2.Dequeue());
            }
            else
            {
                box.Push(c);
            }
        }
        string result = "";
        while(box.Count!= 0)
            result = box.Pop() + result;
        return result;
    }
}
*/
/*
public class Solution {


    public string ReverseParentheses(string s) {
        
        int lb = -1;
        int rb = -1;
        int lb_count = 0;
        for(int i = 0; i < s.Length; i++){
            if (s[i] == '('){
                if (lb == -1) {lb = i;}
                else {lb_count += 1;}
            }
            if (s[i] == ')'){
                if (lb_count == 0) {rb = i; break;}
                else {lb_count -= 1;}
            }
        }
        if (lb == -1 || rb == -1) return s;

        string left_part = "";
        string right_part = "";

        if (lb > 0) {left_part = s.Substring(0, lb);}
        if (rb <= s.Length) {right_part = s.Substring(rb+1);}

        string middle = ReverseParentheses(s.Substring(lb+1, rb-lb-1));
        string rev_mid = new string(middle.Reverse().ToArray());
        
        string res = left_part + rev_mid + ReverseParentheses(right_part);

        return res;
    }
}
*/